### PR TITLE
Feature/table multi select

### LINKED ISSUE #2281: [Feature] Multiselect Table widget should have a select all button
## Summary
Select all button will help users select all rows in that page

### PR DESCRIPTION
**Pull request by techbhavin**
## Description

> Here we add a multi-select feature in the table widget.

Fixes #2281 

## Type of change
- New feature (non-breaking change which adds functionality)

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [  ] I have commented my code, particularly in hard-to-understand areas
- [  ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [  ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: feature/table-multi-select 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 54.15 **(-0.01)** | 36.1 **(-0.02)** | 32.82 **(-0.03)** | 54.64 **(-0.02)**
 :green_circle: | app/client/src/components/designSystems/appsmith/TableComponent/TableStyledWrappers.tsx | 36.05 **(0.68)** | 16.42 **(1.04)** | 0 **(0)** | 36.05 **(0.68)**
 :red_circle: | app/client/src/components/designSystems/appsmith/TableComponent/TableUtilities.tsx | 38.05 **(0.19)** | 18.31 **(-3)** | 6.45 **(-0.96)** | 35.92 **(-0.54)**
 :red_circle: | app/client/src/widgets/TableWidget/index.tsx | 9.21 **(-0.12)** | 0 **(0)** | 10.2 **(-0.44)** | 9.86 **(-0.14)**</details>

### CODE REVIEW

**Comment by somangshu** [comment]
@techbhavin this looks great 👍 

Some alignment issue with the checkbox icon.

**Review by somangshu — DISMISSED** [review]
Approving to run tests

File: app/client/src/components/designSystems/appsmith/TableComponent/TableStyledWrappers.tsx [hunk 1]
```diff
@@ -425,6 +416,27 @@ export const CellWrapper = styled.div<{
...
 
+export const CellCheckboxWrapper = styled(CellWrapper)`
+  justify-content: center;
+  width: 40px;
+`;
+
+export const CellCheckbox = styled.div<{ isChecked: boolean }>`
+  height: 15px;
+  width: 15px;
+  border: 0.5px solid #e7e7e7;
```
Comment (SatishGandham) [inline app/client/src/components/designSystems/appsmith/TableComponent/TableStyledWrappers.tsx:427]: See if you can pick the color from theme.

File: app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx [hunk 2]
```diff
@@ -161,6 +173,30 @@ export function Table(props: TableProps) {
   const tableWrapperRef = useRef<HTMLDivElement | null>(null);
   const tableBodyRef = useRef<HTMLDivElement | null>(null);
   const tableHeaderWrapperRef = React.createRef<HTMLDivElement>();
+  const rowSelectionState = React.useMemo(() => {
+    if (!props.multiRowSelection) return null;
+    // if all rows are un-selected
+    if (every(subPage, (row) => !selectedRowIndices.includes(row.index))) {
```
Comment (SatishGandham) [inline app/client/src/components/designSystems/appsmith/TableComponent/Table.tsx:179]: This can be done without running the loop twice.

**Review by SatishGandham — CHANGES_REQUESTED** [review]
Please check the review comments and tick and - in checkboxes are not centre aligned please.

**Review by SatishGandham — DISMISSED** [review]
Checkbox alignment looks good to.

**Comment by somangshu** [comment]
Re-ran all jobs because flaky tests are failing

**Review by somangshu — APPROVED** [review]

**Comment by somangshu** [comment]
@techbhavin kindly clear all merge conflicts here

**Review by somangshu — APPROVED** [review]

**Comment by YogeshJayaseelan** [comment]
@techbhavin We have tested this feature in deploy preview and please find our observations below 

1) Hovering on Checkbox is not highlighted 

https://loom.com/share/64a1fa3d5ef743f892b056140686346a

2) Checkbox is displayed for empty rows

https://loom.com/share/bea85813986f4d0281a3ec91ac1daba7

**Comment by somangshu** [comment]
@techbhavin Kindly also check the failing tests with resize and hide column.

**Comment by vicky-primathon** [comment]
Sure @somangshu

**Review by somangshu — DISMISSED** [review]

**Review by somangshu — DISMISSED** [review]

**Comment by somangshu** [comment]
@techbhavin can you check the related test failing?

**Review by somangshu — APPROVED** [review]
Approving to re-run tests after fix

**Comment by somangshu** [comment]
Rerunning jobs > `LeaveOrganizationTest_spec.js` is now failing. Unrelated

**Comment by YogeshJayaseelan** [comment]
@techbhavin Tested this feature and still the issue exists - 1)Hovering on Checkbox is not highlighted - https://loom.com/share/64a1fa3d5ef743f892b056140686346a

**Comment by YogeshJayaseelan** [comment]
@somangshu @techbhavin closing this as the issue 1 is expected behaviour

**Pull request merged**